src/dbnode/persist/fs/seek_manager_test.go

```diff
@@ -238,153 +239,175 @@ func TestSeekerManagerBorrowOpenSeekersLazy(t *testing.T) {
+			},
+			{
+				title: "Borrow a seeker from each shard and then modify the clock such that they're out of retention",
+				step: func(m *seekerManager, state *testSeekerManagerOpenCloseLoopState) {
+					blockStartNano := state.startTime.
+						Truncate(state.nsMetadata.Options().RetentionOptions().BlockSize())
+					for _, shard := range state.shards {
+						seeker, err := m.Borrow(shard, blockStartNano)
+						require.NoError(t, err)
+						require.NotNil(t, seeker)
+						state.borrowedSeekers = append(state.borrowedSeekers, seeker)
+					}
+					state.clock.currentTime = state.clock.currentTime.
+						Add((10 * state.nsMetadata.Options().RetentionOptions().RetentionPeriod()))
+				},
+			},
+			{
+				title: "Make sure the seeker manager cant be closed while seekers are borrowed",
+				step: func(m *seekerManager, state *testSeekerManagerOpenCloseLoopState) {
+					require.Equal(t, errCantCloseSeekerManagerWhileSeekersAreBorrowed, m.Close())
+				},
+			},
+			{
+				title: "Make sure that none of the seekers were cleaned up during the openCloseLoop tick (because they're still borrowed)",
+				step: func(m *seekerManager, state *testSeekerManagerOpenCloseLoopState) {
+					m.RLock()
+					blockStartNano := xtime.ToUnixNano(state.startTime.
+						Truncate(state.nsMetadata.Options().RetentionOptions().BlockSize()))
+					for _, shard := range state.shards {
+						require.Equal(t, defaultTestingFetchConcurrency,
+							len(m.seekersByTime(shard).seekers[blockStartNano].active.seekers))
+					}
+					m.RUnlock()
+				},
+			},
+			{
+				title: "Return the borrowed seekers",
+				step: func(m *seekerManager, state *testSeekerManagerOpenCloseLoopState) {
+					blockStartNano := state.startTime.
+						Truncate(state.nsMetadata.Options().RetentionOptions().BlockSize())
+					for i, seeker := range state.borrowedSeekers {
+						require.NoError(t, m.Return(state.shards[i], blockStartNano, seeker))
+					}
+				},
+			},
+			{
+				title: "Make sure that the returned seekers were cleaned up during the openCloseLoop tick",
+				step: func(m *seekerManager, state *testSeekerManagerOpenCloseLoopState) {
+					m.RLock()
+					blockStartNano := xtime.ToUnixNano(state.startTime.
+						Truncate(state.nsMetadata.Options().RetentionOptions().BlockSize()))
+					for _, shard := range state.shards {
+						byTime := m.seekersByTime(shard)
+						byTime.RLock()
+						_, ok := byTime.seekers[blockStartNano]
+						byTime.RUnlock()
+						require.False(t, ok)
+					}
+					m.RUnlock()
+				},
+			},
+		},
+	}
+
+	testSeekerManagerOpenCloseLoop(t, testCase)
+}
+
 // TestSeekerManagerOpenCloseLoop tests the openCloseLoop of the SeekerManager
 // by making sure that it makes the right decisions with regards to cleaning
 // up resources based on their state.
-func TestSeekerManagerOpenCloseLoop(t *testing.T) {
-	defer leaktest.CheckTimeout(t, 1*time.Minute)()
+func testSeekerManagerOpenCloseLoop(t *testing.T, testCase testSeekerManagerOpenCloseLoopCase) {
```

Comment:
So this is debatable, but in recent years I've started to shy away from this form of test control flow (pass a test case to a function) back towards simple setup/teardown. Here's a pattern I use in Go in particular (naming isn't necessarily standard outside of my own code, but the general idea shows up in various places):

```
type testContext struct {
   // contains all dependencies
}

setup := func(t *testing.T) (testContext, func()) {
 // construct dependencies

 return testContext{...}, func() {
     // do teardown
  }
}
```

In this case I think it might simplify things slightly for the reader, without significant cost in terms of verbosity. Could do:

```
func TestSeekerManagerOpenCloseLoop(t *testing.T) {

   t.Run("case1", func(t *testing.T) {
       tctx, teardown := setup(t)
       defer teardown()
 
       executeSteps([]func() {
          ... // same as before
       })
   })
}
```

Thoughts? Not that significantly different I don't think; just makes the control flow slightly simpler.

Reply:
Agreed 👍does improve readability by quite a bit.

Reply:
Merged this PR with other small PRs and fixed this here - https://github.com/m3db/m3/pull/2316